pkg/git/poll/handler.go

```diff
@@ -24,70 +32,96 @@ type Watcher interface {
+	scheduler := quartz.NewStdScheduler()
+	scheduler.Start(ctx)
 	return &Handler{
-		client:      client,
-		watches:     make(map[string]Watcher),
-		createWatch: NewWatch,
-		log:         ctrl.Log.WithName("git-latest-commit-poll-handler"),
+		client:    client,
+		log:       ctrl.Log.WithName("git-latest-commit-poll-handler"),
+		scheduler: scheduler,
+		fetcher:   &git.Fetch{},
 	}
 }
 
-// AddOrModifyGitRepoWatch adds a new watch for the gitrepo if no watch was already present.
-// It updates the existing watch for this gitrepo if present.
-func (h *Handler) AddOrModifyGitRepoWatch(ctx context.Context, gitRepo v1alpha1.GitRepo) {
-	key := getKey(gitRepo)
-	watch, found := h.watches[key]
-	if !found {
-		h.watches[key] = h.createWatch(gitRepo, h.client)
-		h.watches[key].StartBackgroundSync(ctx)
+// AddOrModifyGitRepoPollJob adds a new scheduled job for the gitrepo if no job was already present.
+// It updates the existing job for this gitrepo if present.
+func (h *Handler) AddOrModifyGitRepoPollJob(ctx context.Context, gitRepo v1alpha1.GitRepo) {
+	gitRepoPollKey := GitRepoPollKey(gitRepo)
+	scheduledJob, err := quartzretry.GetScheduledJobWithRetries(h.scheduler, gitRepoPollKey)
+	if err != nil {
+		// job was not found
+		if gitRepo.Spec.DisableGitPolling {
+			// nothing to do if disableGitPolling is set
+			return
+		}
+		h.scheduleJob(ctx, gitRepoPollKey, gitRepo, true)
 	} else {
-		oldSyncInterval := watch.GetSyncInterval()
-		watch.UpdateGitRepo(gitRepo)
-
-		gitRepoSyncInterval := 0.0
-		if pi := gitRepo.Spec.PollingInterval; pi != nil {
-			gitRepoSyncInterval = pi.Seconds()
+		if gitRepo.Spec.DisableGitPolling {
+			// if polling is disabled, just delete the job from the scheduler
+			err = quartzretry.DeleteScheduledJobWithRetries(h.scheduler, gitRepoPollKey)
+			if err != nil {
+				h.log.Error(err, fmt.Sprintf("Error deleting job %s: %v", gitRepoPollKey.String(), err))
+			}
+			return
 		}
-
-		if oldSyncInterval != gitRepoSyncInterval {
-			watch.Restart(ctx)
+		job := scheduledJob.JobDetail().Job()
+		gitRepoPollJob, ok := job.(*GitRepoPollJob)
+		if !ok {
+			h.log.Error(fmt.Errorf("invalid job"),
+				fmt.Sprintf("Error getting Gitrepo poll job, the scheduled job: %s is not a GitRepoPollJob", job.Description()))
+			return
+		}
+		previousInterval := gitRepoPollJob.GitRepo.Spec.PollingInterval
+		gitRepoPollJob.GitRepo = gitRepo
+		if gitRepo.Spec.PollingInterval != previousInterval {
+			// ignoring the error because we're going to schedule immediately
+			_ = quartzretry.DeleteScheduledJobWithRetries(h.scheduler, gitRepoPollKey)
+			h.scheduleJob(ctx, gitRepoPollKey, gitRepo, true)
 		}
 	}
 }
 
-// CleanUpWatches removes all watches whose gitrepo is not present in the cluster.
-func (h *Handler) CleanUpWatches(ctx context.Context) {
+// CleanUpGitRepoPollJobs removes all poll jobs whose gitrepo is not present in the cluster.
+func (h *Handler) CleanUpGitRepoPollJobs(ctx context.Context) {
```

Comment:
I don't see any tests for this code?

Reply:
yeah, good point, original code didn't have any and forgot to add those., thanks for spotting! 
I've added some tests now